### ISSUE TITLE
Make Pagination Component currentPage reset conditional when number of pages changes

### ISSUE THREAD

**Issue by jaylittle** [bootstrap-vue/bootstrap-vue#3716]
Regarding the Pagination component: Would it be possible to make the currentPage reset in the numberOfPages watch conditional using the following logic?

this.currentPage = (newVal >= this.currentPage ? this.currentPage : 1);

The reason being is that I have a situation in which I am using the pagination controls but I have no way of knowing what the total number of rows is in advance. So as the user goes to the next page, I am increasing the total row count based on whether or not the previous page returned a full page of results or not. So this means that when the page size is 25, total rows is set to 26 when we render page 1. When the user requests page 2, we acquire the data and set total rows to 51 if page 2 returned a full 25 rows (incrementing the total number of available pages) or set it to 50 if it did not (which keeps the total number of pages at 2). In the first situation, when data for page 3 is available the pagination control is now changing the page number back to 1 and I have no way to prevent this within code.

TLDR: I basically don't want currentPage to reset back to 1 unless the currentPage value has actually been invalidated by the change in the total number of pages (e.g. currentPage > newTotalNumberOfPages).  Alternatively if we could just add a setting on the component that disables page number resetting, that would also work.

Thanks.

**Comment by tmorehouse** [comment]
We would also need to check if `per-page` (for `b-pagination`) has changed as well, to account for page size changes, and try and keep the user on the new page which had the items they were looking at.

Since the underlying logic is shared with both `b-pagination` and `b-pagination-nav`, we just need to be carefull not to make it introduce bugs in either pagination component.

**Comment by jaylittle** [comment]
That makes sense.  Would it be possible to just add a component property that allows us to disable automatic current page number updates entirely?  That would effectively solve my problem.

**Comment by tmorehouse** [comment]
That might require a bit of additional logic as it could end up (if not used correctly) an un-usable user interface

**Comment by jaylittle** [comment]
That's fair.  It's worth emphasizing that in this case, page size hasn't changed.  Only the total number of rows has changed.

**Comment by tmorehouse** [comment]
But it is possible for total pages to change because of per-page

**Comment by jaylittle** [comment]
Perhaps the solution is to simply have two separate watchers instead of a single watcher on numberOfPages then.  One for perPage and one for totalRows.  perPage watcher can modify currentPage in whatever way it feels is appropriate whereas totalRows watcher maybe doesn't?

**Comment by jaylittle** [comment]
Actually maybe we just need a new watcher for perPage that handles the update for currentPage:

watch: {
  numberOfPages(newVal) {
    if (newVal === this.localNumPages) {
      /* istanbul ignore next */
      return
    }
    this.localNumPages = newVal
  },
  perPage(newVal, oldval) {
    //TODO Handle preservation of current user view
    this.currentPage = 1
  }
},

**Comment by tmorehouse** [comment]
It could be something like that... but will require a bit of testing for various scenarios for both `b-pagination` and `b-pagination-nav` (which perPage is always 1, or doesn't exist)

**Comment by jaylittle** [comment]
I guess one way or another that commit is destined to close this issue, eh?

**Comment by tmorehouse** [comment]
LOL... yeah... darned github automation...

**Comment by tmorehouse** [comment]
BootstrapVue v2.0.0 stable has been released. See https://bootstrap-vue.js.org/docs/misc/changelog
